File: 01_GOF23/DP_examples/Go/01_Creational/Prototype/Product.go
package Prototype

import "fmt"

type Color int

const (
	Unknown Color = iota
	Red
	Green
	Blue
	Yellow
)

// 实现 Color 的 toString 方法
func (c Color) String() string {
	switch c {
	case Red:
		return "Red"
	case Green:
		return "Green"
	case Blue:
		return "Blue"
	case Yellow:
		return "Yellow"
	default:
		return "Unknown"
	}
}

type IShape interface {
	Name() string
	Draw()
	Fill(color Color)
	Clone() IShape
}

type ShapeKind = int

const (
	UNKNOWN_SHAPE ShapeKind = iota
	CIRCLE
	SQUARE
	RECTANGLE
)

type Shape struct {
	id   int
	name string
}

func (s *Shape) Name() string { return s.name }

func (s *Shape) Draw() {}

func (s *Shape) Fill(color Color) {}

type Circle struct {
	Shape
}

func NewCircle(id int, name string) *Circle {
	return &Circle{
		Shape: Shape{
			id:   id,
			name: name,
		},
	}
}
func (c *Circle) Draw() {
	fmt.Println("Draw a Circle")
}

func (c *Circle) Fill(color Color) {
	fmt.Println("Fill with", color)
}

func (c *Circle) Clone() IShape {
	return &Circle{
		Shape: Shape{
			id:   c.id,
			name: c.name,
		},
	}
}

type Square struct {
	Shape
}

func NewSquare(id int, name string) *Square {
	return &Square{
		Shape: Shape{
			id:   id,
			name: name,
		},
	}
}

func (s *Square) Draw() {
	fmt.Println("Draw a Square")
}

func (s *Square) Fill(color Color) {
	fmt.Println("Fill with", color)
}

func (s *Square) Clone() IShape {
	return &Square{
		Shape: Shape{
			id:   s.id,
			name: s.name,
		},
	}
}

type Rectangle struct {
	Shape
}

func NewRectangle(id int, name string) *Rectangle {
	return &Rectangle{
		Shape: Shape{
			id:   id,
			name: name,
		},
	}
}

func (r *Rectangle) Draw() {
	fmt.Println("Draw a Rectangle")
}

func (r *Rectangle) Fill(color Color) {
	fmt.Println("Fill with", color)
}

func (r *Rectangle) Clone() IShape {
	return &Rectangle{
		Shape: Shape{
			id:   r.id,
			name: r.name,
		},
	}
}

File: 01_GOF23/DP_examples/CSharp/01_Creational/Prototype/Product.cs

namespace Creational.Prototype;

public enum Color {
    Unknown,
    Red,
    Green,
    Blue,
    Yellow,
}


public interface IShape {
    string Name { get; }
    void Draw();
    void Fill(Color kind);
}

// 缺省
[Serializable]
public class Shape(int id = 0, string name = "") : IShape {
    public int ID => id;
    public string Name => name;
    // default
    public virtual void Draw() { }
    public virtual void Fill(Color kind) { }
    public virtual T Clone<T>() where T : Shape => MemberwiseClone() as T;
}

// different products
public enum ShapeKind {
    Unknown,
    Circle,
    Square,
    Rectangle
}
[Serializable]
public class Circle : Shape {
    // maybe contain other datas
    public Circle(int id = (int)ShapeKind.Circle, string type = "Circle") : base(id, type) { }
    public override void Draw() => Console.WriteLine("Draw a Circle");
    public override void Fill(Color color) => Console.WriteLine("Fill with " + color.ToString());
}
[Serializable]
public class Square : Shape {
    public Square(int id = (int)ShapeKind.Square, string type = "Square") : base(id, type) { }
    public override void Draw() => Console.WriteLine("Draw a Square");
    public override void Fill(Color color) => Console.WriteLine("Fill with " + color.ToString());
}
[Serializable]
public class Rectangle : Shape {
    public Rectangle(int id = (int)ShapeKind.Rectangle, string type = "Rectangle") : base(id, type) { }
    public override void Draw() => Console.WriteLine("Draw a Rectangle");
    public override void Fill(Color color) => Console.WriteLine("Fill with " + color.ToString());
}
